augur/parse.py

```diff
@@ -162,11 +163,12 @@ def run(args):
     # field to index the dictionary and the data frame
     meta_data = {}
 
-    if 'name' in args.fields:
-        strain_key = 'name'
-    elif 'strain' in args.fields:
-        strain_key = 'strain'
-    else:
+    strain_key = None
+    for possible_id in DEFAULT_ID_COLUMNS:
+        if possible_id in args.fields:
+            strain_key = possible_id
+            break
+    if not strain_key:
```

Comment:
I like the idea of 8dc06946e3fb045d84f03908c9941e965b22daaf but I wonder if there are some workflows relying on the existing order. If that's the case, this would be immediately breaking with no transition period. I think the order should be:

1. (this PR) Add the `--output-id-field` option with a deprecation warning that the order will change from `name`/`strain` to `strain`/`name` and users who wish to keep using `name` should use `--output-id-field name`.
2. Release Augur.
3. Update any workflows that rely on the current default to use `--output-id-field name`.
4. Make another PR to change the default order.
5. Release major version of Augur.

Reply:
Point 1 is a reasonable concern 😄.  I will look into it, at least to understand how to throw a "deprecation warning" in Augur.

For point 3, I am struggling to locate any workflows in nextstrain using the `name` field instead of `strain`. So far, I have only identified workflows that rely on `strain`, and they have not encountered any issues since we do not list a `name` field. But it's quite possible I'm missing something.

https://github.com/search?q=org%3Anextstrain%20%22augur%20parse%22&type=code

Are you finding any?

Reply:
I did the same search but didn't look too deep, since some `--fields` are parameterized which requires extra digging in the Snakemake files.

There are also external and unsearchable usages of Augur which shouldn't be discounted. Although it's possible this behavior isn't being used in practice or the deprecation warning won't actually be read, it's also possible that some workflow out there is relying on this behavior, and it's easy on us to properly handle a change in behavior.

Incidents like #1373 which arise from breaking changes make me super cautious about this stuff now. In that example, at least Biopython got it right by providing deprecation warnings and transition periods indicating which versions support what behavior. We should do something similar here.

Reply:
That makes sense! I made an attempt at a deprecation warning here: https://github.com/nextstrain/augur/pull/1403/commits/40e9a00403318efd3f622fd82ad2e164f7bd7740
But definitely let me know if I was supposed to [import and reuse a "deprecated/warning" function](https://github.com/nextstrain/augur/blob/abc86a81304c2563b1ca307543c1e3cfb74abf83/augur/export_v2.py#L28-L31) somewhere.

Reply:
What you have is fine. The logging functions in export_v2 that you linked are nice, and we should probably move them somewhere central and use them across Augur, but that's extra work that hasn't been done...